statsmodels/tsa/statespace/tools.py

```diff
@@ -0,0 +1,382 @@
+"""
+Statespace Tools
+
+Author: Chad Fulton
+License: Simplified-BSD
+"""
+from __future__ import division, absolute_import, print_function
+
+import numpy as np
+from statsmodels.tools.data import _is_using_pandas
+from . import _statespace
+
+try:
+    from scipy.linalg.blas import find_best_blas_type
+except ImportError:  # pragma: no cover
+    # Shim for SciPy 0.11, derived from tag=0.11 scipy.linalg.blas
+    _type_conv = {'f': 's', 'd': 'd', 'F': 'c', 'D': 'z', 'G': 'z'}
+
+    def find_best_blas_type(arrays):
+        dtype, index = max(
+            [(ar.dtype, i) for i, ar in enumerate(arrays)])
+        prefix = _type_conv.get(dtype.char, 'd')
+        return (prefix, dtype, None)
+
+
+prefix_dtype_map = {
+    's': np.float32, 'd': np.float64, 'c': np.complex64, 'z': np.complex128
+}
+prefix_statespace_map = {
+    's': _statespace.sStatespace, 'd': _statespace.dStatespace,
+    'c': _statespace.cStatespace, 'z': _statespace.zStatespace
+}
+prefix_kalman_filter_map = {
+    's': _statespace.sKalmanFilter, 'd': _statespace.dKalmanFilter,
+    'c': _statespace.cKalmanFilter, 'z': _statespace.zKalmanFilter
+}
+
+
+def companion_matrix(polynomial):
+    r"""
+    Create a companion matrix
+
+    Parameters
+    ----------
+    polynomial : array_like, optional.
+        If an iterable, interpreted as the coefficients of the polynomial from
+        which to form the companion matrix. Polynomial coefficients are in
+        order of increasing degree. If an integer, the size of the companion
+        matrix (the polynomial coefficients are then set to zeros).
+
+    Returns
+    -------
+    companion_matrix : array
+
+    Notes
+    -----
+
+    Returns a matrix of the form
+
+    .. math::
+        \begin{bmatrix}
+            \phi_1 & 1      & 0 & \cdots & 0 \\
+            \phi_2 & 0      & 1 &        & 0 \\
+            \vdots &        &   & \ddots & 0 \\
+                   &        &   &        & 1 \\
+            \phi_n & 0      & 0 & \cdots & 0 \\
+        \end{bmatrix}
+
+    where some or all of the :math:`\phi_i` may be non-zero (if `polynomial` is
+    None, then all are equal to zero).
+
+    If the coefficients provided are :math:`(c_0, c_1, \dots, c_{n})`,
+    then the companion matrix is an :math:`n \times n` matrix formed with the
+    elements in the first column defined as
+    :math:`\phi_i = -\frac{c_i}{c_0}, i \in 1, \dots, n`.
+    """
+    if isinstance(polynomial, int):
+        n = polynomial
+        polynomial = None
+    else:
+        n = len(polynomial) - 1
+        polynomial = np.asanyarray(polynomial)
+
+    matrix = np.zeros((n, n))
+    idx = np.diag_indices(n - 1)
+    idx = (idx[0], idx[1] + 1)
+    matrix[idx] = 1
+    if polynomial is not None and n > 0:
+        matrix[:, 0] = -polynomial[1:] / polynomial[0]
+    return matrix
+
+
+def diff(series, diff=1, seasonal_diff=None, k_seasons=1):
```

Comment:
Having an argument with the same name as the function is bad style, and prevents certain operations like recursion).